### ISSUE TITLE
defined operator of preprocessor

### ISSUE THREAD

**Issue by guwirth** [SonarOpenCommunity/sonar-cxx#923]
@Bertk: The grammer for `defined` operator has to be checked because it shall only be used for #if or #elif directive (see https://msdn.microsoft.com/en-us/library/ew2hz0yd(VS.80).aspx). This limitation is not implemented in current preprocessor grammar.
- preprocessor keywords: https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-squid/src/main/java/org/sonar/cxx/api/CppKeyword.java
- preprocessor grammar with `defined`: https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-squid/src/main/java/org/sonar/cxx/preprocessor/CppGrammar.java#L312
- found also some unit tests: https://github.com/SonarOpenCommunity/sonar-cxx/blob/a227c90869df67023deeecbb3e43148d6eb7a3ae/cxx-squid/src/test/java/org/sonar/cxx/preprocessor/CppGrammarTest.java#L336

**Comment by Bertk** [comment]
This code snippet uses an invalid directive which is accepted by cxx-preprocessor:

```cpp
// test.cpp : Defines the entry point for the console application.
//
 
#include "stdafx.h"
 
#define TEST defined EXAMPLE_H
 
#define defined 1
 
int main()
{
return 0;
}
```
 
Visual Studio throws this warning:

```1>d:\workspaces\test\test\test.cpp(8): warning C4117: macro name 'defined' is reserved, '#defined' ignored.```

**Comment by guwirth** [comment]
@Bertk to be more tolerant is not really an issue?
* Do you wanna see the warning only?
* Or that define is ignored?
* Or both?

**Comment by Bertk** [comment]
@guwirth Creating a warning should be fine for this invalid directive.

**Comment by guwirth** [comment]
@Bertk think we are stating somewhere in our docu that input for plugin should be checked with a compiler and have to be error free.

**Comment by guwirth** [comment]
It is more, not only **defined**:

16.8 Predefined macro names
...
If any of the pre-defined macro names in this subclause, or the identifier defined, is the subject of a #define or a #undef preprocessing directive, the behavior is undefined. Any other predefined macro names shall begin with a leading underscore followed by an uppercase letter or a second underscore.

**Comment by guwirth** [comment]
The compiler is the better place to detect such things. 
Precondition for this plugin: syntactically correct code.
